src/hooks/useTokenInfo.ts

```diff
@@ -0,0 +1,145 @@
+import { useEffect, useState } from 'react'
+import { getERC20TokenContract } from '~/getters'
+import { getPublicWeb3Provider } from '~/shared/stores/wallet'
+
+interface TokenInfo {
+    symbol: string
+    decimals: number
+    name: string
+}
+
+function entryKey(tokenAddress: string, chainId: number) {
+    return `${chainId}-${tokenAddress}`
+}
+
+const tokenInfoCache: Partial<
+    Record<string, TokenInfo | null | Promise<TokenInfo | null>>
+> = {}
+
+export async function getTokenInfo(
+    tokenAddress: string | undefined,
+    chainId: number | undefined,
+): Promise<TokenInfo | null> {
+    if (!tokenAddress || typeof chainId === 'undefined') {
+        return null
+    }
+
+    const key = entryKey(tokenAddress, chainId)
+
+    const current = tokenInfoCache[key]
+
+    if (typeof current === 'undefined') {
+        const contract = getERC20TokenContract({
+            tokenAddress,
+            signer: getPublicWeb3Provider(chainId),
+        })
+
+        const promise = new Promise<TokenInfo | null>((resolve, reject) => {
+            setTimeout(async () => {
+                try {
+                    const symbol = await contract.symbol()
+
+                    if (!symbol) {
+                        throw new Error('Not an ERC-20 token')
+                    }
+
+                    const name = await contract.name()
+
+                    const decimals = await contract.decimals()
+
+                    resolve({
+                        decimals,
+                        name,
+                        symbol,
+                    })
+                } catch (e) {
+                    reject(e)
+                }
+            })
+        })
+
+        tokenInfoCache[key] = promise
+
+        setTimeout(async () => {
+            let next: TokenInfo | null = null
+
+            try {
+                next = await promise
+            } catch (e) {
+                /**
+                 * If we get a rejection from the above promise we simply
+                 * cache `null`. Not the bestest of ways, but we can
+                 * improve as we go forward.
+                 */
+            }
+
+            tokenInfoCache[key] = next
+        })
+
+        return promise
+    }
+
+    return current
```

Comment:
I must say that I kinda don't like the concept that this function may return a value or a promise of this value cause if I'd need to use this for an instance in a service then on each use I'd need to write the logic for handling of the case when it's a promise and when it's a value. What I'd suggest is to always return a promise from this getter. Something like this:
```
return current instanceof Promise ? current : new Promise(resolve => resolve(current))
```

Reply:
It's an `async` getter. It always returns `Promise<TokenInfo | null>`. It's pretty cool. Check its signature!

```ts
export async function getTokenInfo(/* … */): Promise<TokenInfo | null> { /* … */ }
```